### ISSUE TITLE
wxFont::GetPixelSize() incorrectly changes the font size in points

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#23144]
While debugging #23142 I've discovered that doing the following
```cpp
wxFont font1{wxFontInfo{10.5}};
wxFont font2{font1};
font2.GetPixelSize(); // ignore the return value
CHECK( font2.GetFractionalPointSize() == 10.5 );
```
fails because `GetPixelSize()` calls `wxFontRefData::OSXGetCTFont()` which, for some reason, calls `AllocIfNeeded()` which in turn calls `Alloc()` which recreates the font from `CTFontRef` which uses 11pt as size.

At least 2 things seem to be wrong here:

1. Calling `GetPixelSize()` shouldn't recreate anything and I don't understand why does `OSXGetCTFont()` do it. It should be fine to allocate `CTFontRef` for shared font data, after all it's supposed to represent the same font. In fact, not doing it seems to completely negate the whole idea beyond sharing `wxFontRefData` between the fonts in the first place.
2. `CTFontRef` should have 10.5 size and not 11. I have no idea why/how does this happen however, so I don't know how to fix this.

@csomor I'd appreciate your feedback on this, i.e. whether you agree with the proposed change for (1) and if you have any idea about (2). TIA!

**Comment by vadz** [comment]
For the record, calling `GetPixelSize()` changes not just the font size but the font family too: wxOSX is already different from the other ports in that fonts created with `wxFONTFAMILY_DEFAULT` return it from `wxFont::GetFamily()`, so that we need a special workaround for it, see

https://github.com/wxWidgets/wxWidgets/blob/a4f939aafa8f51a6d0ad16b4634387500340976f/src/common/gdicmn.cpp#L789-L798

but after calling `GetPixelSize()`, `GetFamily()` actually starts returning `wxFONTFAMILY_SWISS`, as in the other ports.

I think we should make `GetFamily()` under Mac always return `wxFONTFAMILY_SWISS` too, but at any rate calling `GetPixelSize()` shouldn't modify the font...

**Comment by csomor** [comment]
IIRC: Since truly creating a font meant loosing all former font info properties for which there is no matching font available on the system deferring the creation was needed for passing the test suite.

**Comment by csomor** [comment]
@vadz I'll look at the sample code and try to remember ...

**Comment by vadz** [comment]
> IIRC: Since truly creating a font meant loosing all former font info properties for which there is no matching font available on the system deferring the creation was needed for passing the test suite.

Thinking more about it, it's certainly true that if using the font changes the result of its accessors, we have a problem in either case. I.e. modifying my snippet above slightly:
```cpp
wxFont font1{wxFontInfo{10.5}};
wxFont font2{font1};
font2.GetPixelSize(); // ignore the return value
CHECK( font1.GetFractionalPointSize() == 10.5 ); // Testing the first font now
```
it would be arguably even worse if this check failed.

But, frankly, the current situation is already pretty bad, I pity an unsuspecting user who runs into it because it took even me some time to understand what was going on, you really don't expect using the font to change its size.

If we can't preserve the size (and the family) of the font, we should probably realize it as soon as the first accessor is called then. I.e. if we really can't create fonts of size 10.5 under Mac, it would be better if *all* calls to `GetFractionalPointSize()` returned 11, instead of the current situation when it returns 10.5 initially but then starts retuning 11 later.

Of course, always returning 10.5 would be even better, but if we have to choose, having consistent return value is clearly preferable.

**Comment by csomor** [comment]
I've fixed the fractional point part. Preserving size and family should not be a problem ever. Other properties might be, if I cannot emulate them ...

**Comment by csomor** [comment]
@vadz now the font family issues remain:

#### not using the same font for `wxFONTFAMILY_DEFAULT` and `wxFONTFAMILY_SWISS`
On macOS I'm interpreting `wxFONTFAMILY_DEFAULT` to be the default GUI font, while `wxFONTFAMILY_SWISS` resolves to Helvetica. But perhaps this is overly cautious and people interested in the GUI font start with a `wxSYS_DEFAULT_GUI_FONT` anyway? 

#### staying at `wxFONTFAMILY_DEFAULT`
perhaps this could be fixed at the same place GTK does the conversion:
```
wxFontRefData::wxFontRefData(const wxFontInfo& info)
{
(...)
    else
    {
        wxFontFamily family = info.GetFamily();
        if (family == wxFONTFAMILY_DEFAULT)
            family = wxFONTFAMILY_SWISS;
        SetFamily(family);
    }
```

**Comment by vadz** [comment]
As the actual font selected for `wxFONTFAMILY_DEFAULT` is a "Swiss" (i.e. sans serif[^1]) font, I think we must ensure that its `GetFamily()` returns `wxFONTFAMILY_SWISS` after its creation. The exact way doesn't matter, if you already see one that you like, let's do it like this, if not I can probably find some way to do it. The important question is whether you agree with doing this?

[^1]: I think things would be much more clear if we renamed our font families constants and used `SERIF` instead of `ROMAN` and `SANS_SERIF` instead of `SWISS`. And also `FIXED` instead of `MODERN` which makes the least sense of all of them.

**Comment by vadz** [comment]
Everything here seems to be fixed now, so closing, thanks again!

I've also backported the changes done to fix this to 3.2 branch and plan to include them in 3.2.2.
